File: z3wire/weave/resolver.cc
#include "z3wire/weave/resolver.h"

#include <algorithm>
#include <cstdint>
#include <set>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

#include "absl/strings/str_format.h"
#include "z3wire/weave/rdl.pb.h"

namespace z3wire::weave {
namespace {

const std::set<std::string> kReservedNames = {
    "Create", "Pack", "ToConcrete", "FromConcrete", "ToProto", "FromProto",
};

std::vector<ResolvedEnum> ResolveEnums(const z3wire_rdl::Module& module) {
  std::vector<ResolvedEnum> resolved;
  for (const auto& enum_def : module.enums()) {
    uint64_t max_val = (1ULL << enum_def.width()) - 1;
    for (const auto& v : enum_def.values()) {
      if (v.value() > max_val) {
        throw std::invalid_argument(absl::StrFormat(
            "Enum '%s': value '%s' = %d exceeds width %d (max %d)",
            enum_def.name(), v.name(), v.value(), enum_def.width(), max_val));
      }
    }
    std::set<uint64_t> seen_values;
    for (const auto& v : enum_def.values()) {
      if (seen_values.count(v.value())) {
        throw std::invalid_argument(absl::StrFormat(
            "Enum '%s': duplicate value %d", enum_def.name(), v.value()));
      }
      seen_values.insert(v.value());
    }
    ResolvedEnum re;
    re.name = enum_def.name();
    re.desc = enum_def.desc();
    re.width = static_cast<int>(enum_def.width());
    for (const auto& v : enum_def.values()) {
      re.values.push_back({v.name(), v.desc(), v.value()});
    }
    resolved.push_back(std::move(re));
  }
  return resolved;
}

int FieldElementWidth(
    const z3wire_rdl::FieldType& field_type,
    const std::unordered_map<std::string, ResolvedEnum>& enum_map,
    const std::unordered_map<std::string, ResolvedStruct>& struct_map) {
  switch (field_type.kind_case()) {
    case z3wire_rdl::FieldType::kBool:
      return 1;
    case z3wire_rdl::FieldType::kBitvec:
      return static_cast<int>(field_type.bitvec().width());
    case z3wire_rdl::FieldType::kEnumRef: {
      auto it = enum_map.find(field_type.enum_ref());
      if (it == enum_map.end()) {
        throw std::invalid_argument(
            absl::StrFormat("Unknown enum_ref '%s'", field_type.enum_ref()));
      }
      return it->second.width;
    }
    case z3wire_rdl::FieldType::kStructRef: {
      auto it = struct_map.find(field_type.struct_ref());
      return it->second.total_width;
    }
    default:
      throw std::invalid_argument("Unknown field type");
  }
}

ResolvedStruct::PackOrder ResolvePackOrder(
    const z3wire_rdl::Module& module, const z3wire_rdl::Struct& struct_def) {
  auto order = struct_def.field_pack_order();
  if (order == z3wire_rdl::FIELD_PACK_ORDER_UNSPECIFIED) {
    order = module.field_pack_order();
  }
  if (order == z3wire_rdl::FIELD_PACK_ORDER_UNSPECIFIED) {
    throw std::invalid_argument(absl::StrFormat(
        "Struct '%s': field_pack_order not specified at struct or module level",
        struct_def.name()));
  }
  if (order == z3wire_rdl::FIELD_PACK_ORDER_LSB_FIRST) {
    return ResolvedStruct::kLsbFirst;
  }
  return ResolvedStruct::kMsbFirst;
}

std::vector<ResolvedStruct> ResolveStructs(
    const z3wire_rdl::Module& module,
    const std::unordered_map<std::string, ResolvedEnum>& enum_map) {
  std::unordered_map<std::string, ResolvedStruct> struct_map;
  std::vector<ResolvedStruct> resolved;

  // Topological sort via retry loop.
  std::vector<z3wire_rdl::Struct> pending(module.structs().begin(),
                                          module.structs().end());
  int max_iterations =
      static_cast<int>(pending.size()) * static_cast<int>(pending.size()) + 1;
  int iteration = 0;

  while (!pending.empty()) {
    ++iteration;
    if (iteration > max_iterations) {
      std::string names;
      for (const auto& s : pending) {
        if (!names.empty()) names += ", ";
        names += "'" + s.name() + "'";
      }
      throw std::invalid_argument(
          absl::StrFormat("Circular struct dependency: [%s]", names));
    }

    auto struct_def = std::move(pending.front());
    pending.erase(pending.begin());

    // Check if all struct_ref dependencies are resolved.
    bool deps_met = true;
    for (const auto& f : struct_def.fields()) {
      if (f.type().kind_case() == z3wire_rdl::FieldType::kStructRef &&
          struct_map.find(f.type().struct_ref()) == struct_map.end()) {
        deps_met = false;
        break;
      }
    }
    if (!deps_met) {
      pending.push_back(std::move(struct_def));
      continue;
    }

    auto pack_order = ResolvePackOrder(module, struct_def);

    // Validate field names.
    for (const auto& f : struct_def.fields()) {
      if (kReservedNames.count(f.name())) {
        throw std::invalid_argument(
            absl::StrFormat("Struct '%s': field name '%s' is reserved "
                            "(collides with generated method)",
                            struct_def.name(), f.name()));
      }
      if (f.type().kind_case() == z3wire_rdl::FieldType::kEnumRef &&
          enum_map.find(f.type().enum_ref()) == enum_map.end()) {
        throw std::invalid_argument(
            absl::StrFormat("Struct '%s', field '%s': unknown enum_ref '%s'",
                            struct_def.name(), f.name(), f.type().enum_ref()));
      }
    }

    // First pass: compute field widths.
    struct FieldInfo {
      const z3wire_rdl::Field* field;
      ResolvedField::Kind kind;
      int element_width;
      int count;
      int total_width;
      ResolvedField::Signedness signedness;
    };
    std::vector<FieldInfo> field_infos;

    for (const auto& f : struct_def.fields()) {
      int element_width = FieldElementWidth(f.type(), enum_map, struct_map);
      int count = static_cast<int>(f.count());
      int total_width = element_width * std::max(count, 1);

      ResolvedField::Kind kind;
      switch (f.type().kind_case()) {
        case z3wire_rdl::FieldType::kBool:
          kind = ResolvedField::kBool;
          break;
        case z3wire_rdl::FieldType::kBitvec:
          kind = ResolvedField::kBitVec;
          break;
        case z3wire_rdl::FieldType::kEnumRef:
          kind = ResolvedField::kEnumRef;
          break;
        case z3wire_rdl::FieldType::kStructRef:
          kind = ResolvedField::kStructRef;
          break;
        default:
          kind = ResolvedField::kBitVec;
          break;
      }

      auto signedness = ResolvedField::kUnsigned;
      if (f.type().kind_case() == z3wire_rdl::FieldType::kBitvec &&
          f.type().bitvec().signedness() == z3wire_rdl::SIGNEDNESS_SIGNED) {
        signedness = ResolvedField::kSigned;
      }

      field_infos.push_back(
          {&f, kind, element_width, count, total_width, signedness});
    }

    int struct_total_width = 0;
    for (const auto& fi : field_infos) {
      struct_total_width += fi.total_width;
    }

    // Second pass: assign offsets based on pack order.
    std::vector<ResolvedField> fields;
    if (pack_order == ResolvedStruct::kLsbFirst) {
      int offset = 0;
      for (const auto& fi : field_infos) {
        ResolvedField rf;
        rf.name = fi.field->name();
        rf.desc = fi.field->desc();
        rf.width = fi.total_width;
        rf.offset = offset;
        rf.count = fi.count;
        rf.element_width = fi.element_width;
        rf.reserved = fi.field->reserved();
        rf.kind = fi.kind;
        rf.signedness = fi.signedness;
        if (fi.kind == ResolvedField::kEnumRef) {
          rf.enum_name = fi.field->type().enum_ref();
        }
        if (fi.kind == ResolvedField::kStructRef) {
          rf.struct_name = fi.field->type().struct_ref();
        }
        fields.push_back(std::move(rf));
        offset += fi.total_width;
      }
    } else {
      // MSB_FIRST: first field in list occupies highest bits.
      int offset = struct_total_width;
      for (const auto& fi : field_infos) {
        offset -= fi.total_width;
        ResolvedField rf;
        rf.name = fi.field->name();
        rf.desc = fi.field->desc();
        rf.width = fi.total_width;
        rf.offset = offset;
        rf.count = fi.count;
        rf.element_width = fi.element_width;
        rf.reserved = fi.field->reserved();
        rf.kind = fi.kind;
        rf.signedness = fi.signedness;
        if (fi.kind == ResolvedField::kEnumRef) {
          rf.enum_name = fi.field->type().enum_ref();
        }
        if (fi.kind == ResolvedField::kStructRef) {
          rf.struct_name = fi.field->type().struct_ref();
        }
        fields.push_back(std::move(rf));
      }
    }

    if (struct_def.has_width() &&
        static_cast<int>(struct_def.width()) != struct_total_width) {
      throw std::invalid_argument(absl::StrFormat(
          "Struct '%s': declared width %d but fields sum to %d",
          struct_def.name(), struct_def.width(), struct_total_width));
    }

    ResolvedStruct rs;
    rs.name = struct_def.name();
    rs.desc = struct_def.desc();
    rs.total_width = struct_total_width;
    rs.field_pack_order = pack_order;
    rs.fields = std::move(fields);
    struct_map[rs.name] = rs;
    resolved.push_back(std::move(rs));
  }

  return resolved;
}

}  // namespace

ResolvedModule Resolve(const z3wire_rdl::Module& module) {
  if (module.file_prefix().empty()) {
    throw std::invalid_argument("Module must specify file_prefix");
  }
  auto enums = ResolveEnums(module);
  std::unordered_map<std::string, ResolvedEnum> enum_map;
  for (const auto& e : enums) {
    enum_map[e.name] = e;
  }
  auto structs = ResolveStructs(module, enum_map);
  return ResolvedModule{
      .file_prefix = module.file_prefix(),
      .ns = module.namespace_(),
      .enums = std::move(enums),
      .structs = std::move(structs),
  };
}

}  // namespace z3wire::weave

File: z3wire/weave/resolver.py
"""Resolves an RDL Module: validates, computes widths and bit offsets."""

from dataclasses import dataclass

from z3wire.weave import rdl_pb2

_RESERVED_NAMES = {
    "Create",
    "Pack",
    "ToConcrete",
    "FromConcrete",
    "ToProto",
    "FromProto",
}


@dataclass
class ResolvedField:
    name: str
    desc: str
    width: int  # Total width in bits (including count)
    offset: int  # Bit offset within parent struct
    count: int  # 0 = scalar, >= 1 = array
    element_width: int  # Width of a single element
    reserved: bool
    # Type info
    kind: str  # "bool", "bitvec", "enum_ref", "struct_ref"
    signedness: str  # "unsigned" or "signed" (for bitvec)
    enum_name: str = ""
    struct_name: str = ""


@dataclass
class ResolvedEnum:
    name: str
    desc: str
    width: int
    values: list  # list of (name, desc, value) tuples


@dataclass
class ResolvedStruct:
    name: str
    desc: str
    total_width: int
    field_pack_order: str  # "lsb_first"
    fields: list  # list of ResolvedField


@dataclass
class ResolvedModule:
    file_prefix: str
    namespace: str
    enums: list  # list of ResolvedEnum
    structs: list  # list of ResolvedStruct


def resolve(module: rdl_pb2.Module) -> ResolvedModule:
    """Resolve an RDL module: validate and compute layouts."""
    if not module.file_prefix:
        raise ValueError("Module must specify file_prefix")
    enums = _resolve_enums(module)
    enum_map = {e.name: e for e in enums}
    structs = _resolve_structs(module, enum_map)
    return ResolvedModule(
        file_prefix=module.file_prefix,
        namespace=module.namespace,
        enums=enums,
        structs=structs,
    )


def _resolve_enums(module):
    resolved = []
    for enum_def in module.enums:
        # Validate values fit within width
        max_val = (1 << enum_def.width) - 1
        for v in enum_def.values:
            if v.value > max_val:
                raise ValueError(
                    f"Enum '{enum_def.name}': value '{v.name}' = {v.value} "
                    f"exceeds width {enum_def.width} (max {max_val})"
                )
        # Validate unique values
        seen_values = set()
        for v in enum_def.values:
            if v.value in seen_values:
                raise ValueError(
                    f"Enum '{enum_def.name}': duplicate value {v.value}"
                )
            seen_values.add(v.value)

        values = [(v.name, v.desc, v.value) for v in enum_def.values]
        resolved.append(
            ResolvedEnum(
                name=enum_def.name,
                desc=enum_def.desc,
                width=enum_def.width,
                values=values,
            )
        )
    return resolved


def _field_element_width(field_type, enum_map, struct_map):
    kind = field_type.WhichOneof("kind")
    if kind == "bool":
        return 1
    elif kind == "bitvec":
        return field_type.bitvec.width
    elif kind == "enum_ref":
        if field_type.enum_ref not in enum_map:
            raise ValueError(f"Unknown enum_ref '{field_type.enum_ref}'")
        return enum_map[field_type.enum_ref].width
    elif kind == "struct_ref":
        return struct_map[field_type.struct_ref].total_width
    else:
        raise ValueError(f"Unknown field type: {kind}")


def _resolve_pack_order(module, struct):
    order = struct.field_pack_order
    if order == rdl_pb2.FIELD_PACK_ORDER_UNSPECIFIED:
        order = module.field_pack_order
    if order == rdl_pb2.FIELD_PACK_ORDER_UNSPECIFIED:
        raise ValueError(
            f"Struct '{struct.name}': field_pack_order not specified "
            f"at struct or module level"
        )
    if order == rdl_pb2.FIELD_PACK_ORDER_LSB_FIRST:
        return "lsb_first"
    return "msb_first"


def _resolve_structs(module, enum_map):
    struct_map = {}
    resolved = []

    # Sort structs so dependencies come first.
    pending = list(module.structs)
    max_iterations = len(pending) * len(pending) + 1
    iteration = 0
    while pending:
        iteration += 1
        if iteration > max_iterations:
            names = [s.name for s in pending]
            raise ValueError(f"Circular struct dependency: {names}")
        struct = pending.pop(0)
        # Check if all struct_ref dependencies are resolved
        deps_met = True
        for f in struct.fields:
            kind = f.type.WhichOneof("kind")
            if kind == "struct_ref" and f.type.struct_ref not in struct_map:
                deps_met = False
                break
        if not deps_met:
            pending.append(struct)
            continue

        pack_order = _resolve_pack_order(module, struct)

        # Validate field names
        for f in struct.fields:
            if f.name in _RESERVED_NAMES:
                raise ValueError(
                    f"Struct '{struct.name}': field name '{f.name}' is "
                    f"reserved (collides with generated method)"
                )
            kind = f.type.WhichOneof("kind")
            if kind == "enum_ref" and f.type.enum_ref not in enum_map:
                raise ValueError(
                    f"Struct '{struct.name}', field '{f.name}': "
                    f"unknown enum_ref '{f.type.enum_ref}'"
                )

        # First pass: compute field widths.
        field_infos = []
        for f in struct.fields:
            kind = f.type.WhichOneof("kind")
            element_width = _field_element_width(f.type, enum_map, struct_map)
            count = f.count
            total_width = element_width * max(count, 1)

            signedness = "unsigned"
            if (
                kind == "bitvec"
                and f.type.bitvec.signedness == rdl_pb2.SIGNEDNESS_SIGNED
            ):
                signedness = "signed"

            field_infos.append(
                (f, kind, element_width, count, total_width, signedness)
            )

        struct_total_width = sum(fw for _, _, _, _, fw, _ in field_infos)

        # Second pass: assign offsets based on pack order.
        fields = []
        if pack_order == "lsb_first":
            offset = 0
            for (
                f,
                kind,
                element_width,
                count,
                total_width,
                signedness,
            ) in field_infos:
                fields.append(
                    ResolvedField(
                        name=f.name,
                        desc=f.desc,
                        width=total_width,
                        offset=offset,
                        count=count,
                        element_width=element_width,
                        reserved=f.reserved,
                        kind=kind,
                        signedness=signedness,
                        enum_name=(
                            f.type.enum_ref if kind == "enum_ref" else ""
                        ),
                        struct_name=(
                            f.type.struct_ref if kind == "struct_ref" else ""
                        ),
                    )
                )
                offset += total_width
        else:
            # MSB_FIRST: first field in list occupies highest bits.
            offset = struct_total_width
            for (
                f,
                kind,
                element_width,
                count,
                total_width,
                signedness,
            ) in field_infos:
                offset -= total_width
                fields.append(
                    ResolvedField(
                        name=f.name,
                        desc=f.desc,
                        width=total_width,
                        offset=offset,
                        count=count,
                        element_width=element_width,
                        reserved=f.reserved,
                        kind=kind,
                        signedness=signedness,
                        enum_name=(
                            f.type.enum_ref if kind == "enum_ref" else ""
                        ),
                        struct_name=(
                            f.type.struct_ref if kind == "struct_ref" else ""
                        ),
                    )
                )

        total_width = struct_total_width
        if struct.HasField("width") and struct.width != total_width:
            raise ValueError(
                f"Struct '{struct.name}': declared width {struct.width} "
                f"but fields sum to {total_width}"
            )

        rs = ResolvedStruct(
            name=struct.name,
            desc=struct.desc,
            total_width=total_width,
            field_pack_order=pack_order,
            fields=fields,
        )
        struct_map[struct.name] = rs
        resolved.append(rs)

    return resolved
